src/server/pkg/storage/fileset/compaction.go

```diff
@@ -0,0 +1,168 @@
+package fileset
+
+import (
+	"context"
+	"time"
+
+	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index"
+	"golang.org/x/sync/errgroup"
+)
+
+// IsCompacted returns true if the fileset is already in compacted form.
+func (s *Storage) IsCompacted(ctx context.Context, id ID) (bool, error) {
+	md, err := s.store.Get(ctx, id)
+	if err != nil {
+		return false, err
+	}
+	switch x := md.Value.(type) {
+	case *Metadata_Composite:
+		ids, err := s.Flatten(ctx, x.Composite.Layers)
+		if err != nil {
+			return false, err
+		}
+		layers, err := s.getPrimitiveBatch(ctx, ids)
+		if err != nil {
+			return false, err
+		}
+		return isCompacted(s.levelFactor, layers), nil
+	case *Metadata_Primitive:
+		return true, nil
+	default:
+		// TODO: should it be?
```

Comment:
I guess it would be compacted, but we should probably discuss empty filesets more.

Reply:
There is no way to create an empty fileset through the API, so for the moment I think an error is more useful than considering it trivially compacted.